### ISSUE TITLE
ANSI colors no longer render in cmd.exe

### ISSUE THREAD

**Issue by cowwoc** [Microsoft/BashOnWindows#1173]
Please use the following bug reporting template to help produce actionable and reproducible issues. Please try to ensure that the reproduction is minimal so that the team can go through more bugs!

* A brief description
https://github.com/Microsoft/BashOnWindows/issues/76 and https://github.com/Microsoft/BashOnWindows/issues/345 added support for ANSI colors. This works in version 10.0.10586 but got broken in version 10.0.14393.

https://github.com/symfony/symfony/issues/19520 confirms that other people are seeing the same problem.

* Expected results
ANSI colors should render in cmd.exe when applications output to stdout/stderr.

* Actual results (with terminal output if applicable)
Raw ANSI codes are seen. No colors are rendered.

* Your Windows build number
 10.0.14393

* Steps / All commands required to reproduce the error from a brand new installation 
1. Run this Java testcase:

```
public class Test {

    public static void main(String[] args) {
        System.out.println("\033[37;1mwhite\033[0m");
    }
}
```

2. You will get this output: `[37;1mwhite[0m`
3. Redirect the application stdout to `foo.log` and run `type foo.log`
4. Notice that the output now reads `white` using white text

This is reproducible 100% of the time.

**Comment by fpqc** [comment]
10.0.14393 is not the most current release.  This should work totally fine in 14942 (latest).

**Comment by cowwoc** [comment]
@fpqc https://support.microsoft.com/en-ca/help/12387/windows-10-update-history disagrees with your assertion. I tried Windows Update locally (Windows 10 Pro) and it says I have the latest version.

I assume you're talking about an insider release?

**Comment by fpqc** [comment]
Yes it's talking about an insider release, and yes I've attempted it.
![2016-10-07 2](https://cloud.githubusercontent.com/assets/6083371/19202085/b017e422-8c9e-11e6-82ae-df9a5bd82641.png)

**Comment by Manouchehri** [comment]
@cowwoc For reporting bugs, you really need to be on the fast ring insider builds. The development here is done way faster than the Windows stable builds can keep up.

**Comment by bitcrazed** [comment]
Closing: [24-bit color color support was first introduced in Win10 Insiders Update 14931](https://blogs.msdn.microsoft.com/commandline/2016/09/22/24-bit-color-in-the-windows-console/), and isn't available in the Win10 Anniversary Update.

To exercise this feature, you'll need to [join the Windows Insiders program](https://insider.windows.com/) and update your PC to install Fast-ring releases. Alternatively, you can opt for slow-ring updates and wait a few weeks for this feature to arrive.

**Comment by cowwoc** [comment]
Please reopen this issue. I am seeing the same bug in update 14946.

![image](https://cloud.githubusercontent.com/assets/633348/19404996/ded5bf62-9240-11e6-8380-276a9a38bee0.png)

**Comment by fpqc** [comment]
@cowwoc you have to set the console mode for printf.exe to enable virtual terminal processing, probably by recompiling it.

Ask the gnuwin32 people to enable.

**Comment by cowwoc** [comment]
@fpqc I don't really care about printf. I was just using it because you were doing the same in your testcase. The only thing I care about is getting this to work with Java.

1. I tried the printf version you mentioned and got the same output (no colors).
2. Can you please run *my* testcase on your end?
3. Why isn't ANSI support enabled by default (without having to enable `ENABLE_VIRTUAL_TERMINAL_INPUT`?
4. Why were ANSI colors working for me (and others) in past releases and now they no longer do? This seems like a clear regression.
5. How do I find out what console mode a process is running with? I want to reverse-engineer what Java is using.

**Comment by fpqc** [comment]
@cowwoc No idea, can you set console modes for java.exe?  Request that Sun enables it in jvm?

**Comment by cowwoc** [comment]
@fpqc There is no way to set console modes in Java without using JNI (which is not a realistic option for my use-case). It is also unlikely Sun/Oracle would enable this in the near future.

I think it would help if someone from Microsoft address point 4: why were ANSI colors working in 10.0.10586 but got broken in version 10.0.14393? Something must have changed and I'm not convinced yet this was intentional.

**Comment by fpqc** [comment]
@cowwoc I dunno. I can't help you, and sorry, but the devs on this Issues page can't help you either.  I know Rich Turner (Senior PM on the console enhancement project) has a special work-only Twitter account and maybe you can ask him why the default got changed (vis-a-vis escape sequence processing) and ask if there is a way to turn it on for a console java application.

**Comment by cowwoc** [comment]
@fpqc Perhaps @zadjii-msft could help? He is the one who marked https://github.com/Microsoft/BashOnWindows/issues/1173 as fixed.

**Comment by fpqc** [comment]
@cowwoc I dunno, Mr. Turner is kind of his boss, I think.

**Comment by fpqc** [comment]
@cowwoc Check out the top answer here:

http://stackoverflow.com/questions/12404258/truly-interactive-command-line

There is a project called Java Native Access (JNA) that lets you call SetConsoleMode from Java.

**Comment by cowwoc** [comment]
@fpqc JNA uses JNI under the hood, so this isn't an option. The only way to interact with native code is through JNI.

**Comment by fpqc** [comment]
@cowwoc I guess then you've gotta ask the Java people to set it or write a small wrapper program.

**Comment by fpqc** [comment]
@cowwoc JNA docu says it doesn't use JNI under the hood:

>JNA provides Java programs easy access to native shared libraries without writing anything but Java code - no JNI or native code is required. This functionality is comparable to Windows' Platform/Invoke and Python's ctypes.

**Comment by cowwoc** [comment]
@fpqc They simply mean that you don't need any *extra* JNI code. Quoting https://github.com/java-native-access/jna:

> JNA uses a small JNI library stub to dynamically invoke native code

**Comment by fpqc** [comment]
@cowwoc Yeah that's why I deleted my post.  I continued reading it.

What is this super-restrictive use case you have with no JNI at all? Maybe you can be a hero and write a generic mini-loader for Java, then throw it up on github?

**Comment by cowwoc** [comment]
@bitcrazed Seeing as you closed this issue, can you please clarify why ANSI colors were working in 10.0.10586 but no longer work in 10.0.14393?

@fpqc It's a long story, but the jist of it is that JNI is a deployment nightmare. You have to deploy different JAR files for different platforms. On platforms that use native code, you have to unpack the libraries before loading them. Maven (the prominent build system for Java) handles building and packaging native libraries very poorly. And so on...

The minute I introduce native code to my library I am pushing this complexity on users of my library (and their users, transitively). Considering what my library does (https://bitbucket.org/cowwoc/requirements/) this would be a major deterrent to its use.

I seriously hope that there is a way to use this functionality without invoking `SetConsoleMode()`. I guess we'll have to wait for Monday to find out.

**Comment by fpqc** [comment]
@cowwoc Isn't that the point of JNA, no need to extract all those native libraries?  You just have the single dll, and that lets you make calls to the Windows API, just a single quick branch on initialization?

**Comment by cowwoc** [comment]
@fpqc I appreciate you trying to help, but whether it's one DLL or many, you still end up with the deployment problems I mentioned.

**Comment by Manouchehri** [comment]
Is the problem still appearing if you use ConEmu?

https://conemu.github.io

**Comment by fpqc** [comment]
@manouchehri That's not the problem, he wants his java application to run on unpatched Windows conhost in stock configuration and default settings with color-escape sequence processing enabled.

**Comment by cowwoc** [comment]
@Manouchehri As fpqc mentioned, this works in conemu, but I want to get my Java application to output ANSI colors on an unpatched Windows conhost using the default configuration.

**Comment by Manouchehri** [comment]
It works fine in `cmd.exe` on my end. Cannot reproduce.

![image](https://cloud.githubusercontent.com/assets/7232674/19418148/33a6d5d4-938c-11e6-82b8-35d0e25a3597.png)

```
Microsoft Windows [Version 10.0.14946]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\Users\david>bash
dave@universal:/mnt/c/Users/david$ printf "\033[37;1mwhite\033[0m\n"
white
```

**Comment by fpqc** [comment]
@manouchehri It depends on the console application. Bash has that mode set, but Java doesnt.

**Comment by zadjii-msft** [comment]
Oh my, you leave the office for one weekend, and this happens :P

There's a lot going on in here, so let me try to clarify: The windows console only supports VT sequences of ANY kind with the `ConsoleMode` set to enable Virtual Terminal support. See [This documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686033(v=vs.85).aspx) on `SetConsoleMode`. Specifically, you need to turn on the output mode flag `ENABLE_VIRTUAL_TERMINAL_PROCESSING`.

By default, `cmd.exe` enables VT mode for itself. Now here's the confusion:
 - In **10.0.10586** (or 1511 or TH2), it left this mode on when calling child applications. This meant that anything you launched from cmd would get the new behavior for free. Of course, we found that *unsurprisingly* broke compatibility, so we had to change that. So...
 - In **10.0.14393** (Anniversary update, or greater), cmd uses VT mode, but then reverts it for launching child processes. So if you wrote a console app that uses VT sequences after October 2015, then tried running it on an AU system, you'll find that the sequences don't work anymore. You'll have to manually turn on VT support with `SetConsoleMode` yourself. **IMPORTANT** cmd also wipes out the changes a child makes to the mode. So you can't call `enable_vt.exe` & `myapp.exe` and have the VT mode from `enable_vt.exe` pollute the mode of `myapp.exe`

To further confuse, `bash.exe` turns on VT mode itself, because obviously, it needs that. 

I can't really speak about the Java issue at hand, it's been far too long since I've used Java, and I don't really have any experience calling Win32 API's from java. I did end up writing a snippet of code that enables VT support on Windows in python, so it can be done.  

In conclusion: I'm going to leave this issue closed, as this is the proper, expected behavior. Confusing, but correct.

Also, in terms of hierarchy, @bitcrazed is the PM for the console, while I'm one of the devs

**Comment by cowwoc** [comment]
@zadjii-msft Thanks for the clarification. It helps a lot.

Out of curiosity, can you please provide concrete examples where enabling this flag by default broke app compatibility?

**Comment by WASasquatch** [comment]
Having this issue on the latest Windows 10 Creators. Was working fine (for all my Minecraft Servers) now, the MC servers will error out with 

```
    2017-05-11 10:50:34,157 WARN Unable to instantiate org.fusesource.jansi.WindowsAnsiOutputStream
    [ERROR] Terminal initialization failed; falling back to unsupported
    [STDERR]: java.lang.NoClassDefFoundError: Could not initialize class org.fusesource.jansi.internal.Kernel32
    [STDERR]:       at org.fusesource.jansi.internal.WindowsSupport.getConsoleMode(WindowsSupport.java:50)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.WindowsTerminal.getConsoleMode(WindowsTerminal.java:204)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.WindowsTerminal.init(WindowsTerminal.java:82)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.TerminalFactory.create(TerminalFactory.java:101)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.TerminalFactory.get(TerminalFactory.java:158)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:229)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:221)
    [STDERR]:       at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:213)
    [STDERR]:       at net.minecraft.server.v1_11_R1.MinecraftServer.<init>(MinecraftServer.java:149)
    [STDERR]:       at net.minecraft.server.v1_11_R1.DedicatedServer.<init>(DedicatedServer.java:48)
    [STDERR]:       at net.minecraft.server.v1_11_R1.MinecraftServer.main(MinecraftServer.java:1006)
    [STDERR]:       at org.bukkit.craftbukkit.Main.main(Main.java:226)
    [STDERR]:       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [STDERR]:       at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    [STDERR]:       at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    [STDERR]:       at java.lang.reflect.Method.invoke(Unknown Source)
    [STDERR]:       at com.destroystokyo.paperclip.Paperclip.main(Paperclip.java:239)

    [10:50:38 INFO]: [STDERR]: java.lang.NoClassDefFoundError: Could not initialize class org.fusesource.jansi.internal.Kernel32
    [10:50:38 INFO]: [STDERR]:      at org.fusesource.jansi.internal.WindowsSupport.getConsoleMode(WindowsSupport.java:50)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.WindowsTerminal.getConsoleMode(WindowsTerminal.java:204)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.WindowsTerminal.init(WindowsTerminal.java:82)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.TerminalFactory.create(TerminalFactory.java:101)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.TerminalFactory.get(TerminalFactory.java:158)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:229)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:221)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.libs.jline.console.ConsoleReader.<init>(ConsoleReader.java:213)
    [10:50:38 INFO]: [STDERR]:      at net.minecraft.server.v1_11_R1.MinecraftServer.<init>(MinecraftServer.java:149)
    [10:50:38 INFO]: [STDERR]:      at net.minecraft.server.v1_11_R1.DedicatedServer.<init>(DedicatedServer.java:48)
    [10:50:38 INFO]: [STDERR]:      at net.minecraft.server.v1_11_R1.MinecraftServer.main(MinecraftServer.java:1006)
    [10:50:38 INFO]: [STDERR]:      at org.bukkit.craftbukkit.Main.main(Main.java:226)
    [10:50:38 INFO]: [STDERR]:      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [10:50:38 INFO]: [STDERR]:      at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    [10:50:38 INFO]: [STDERR]:      at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    [10:50:38 INFO]: [STDERR]:      at java.lang.reflect.Method.invoke(Unknown Source)
    [10:50:38 INFO]: [STDERR]:      at com.destroystokyo.paperclip.Paperclip.main(Paperclip.java:239)
```

**Comment by cowwoc** [comment]
@WASasquatch The error you pasted has nothing to do with Windows. Please google `NoClassDefFoundError` and discuss with the [jansi](https://github.com/fusesource/jansi) authors if necessary.

**Comment by WASasquatch** [comment]
Same issue with just printing color codes via any program. I was using this as a program-breaking example.

**Comment by cowwoc** [comment]
@WASasquatch Can you reproduce this issue without using jansi? If so, can you provide a testcase others can run?

**Comment by zadjii-msft** [comment]
@WASasquatch That definitely looks like a java problem, not a conhost or WSL problem to me. 

Do you have a specific, simple example of VT sequences not working in WSL? They're a pretty heavily tested part of the system, I doubt they would have broken:

![image](https://cloud.githubusercontent.com/assets/18356694/25969686/bdc810a0-364a-11e7-9536-287de7027d09.png)
worksonmymachine.png

**Comment by glenn-slayden** [comment]
As noted [here](https://superuser.com/a/1300251/459921) and [here](https://github.com/rg3/youtube-dl/issues/15758), the **ANSI VT** can be enabled for *SetConsoleMode-oblivious* apps, without having to re-compile them, by changing the following registry key.

    [HKEY_CURRENT_USER\Console]
    "VirtualTerminalLevel"=dword:00000001

This changes the system-wide default for ANSI VT processing from ***opt-in*** to ***opt-out***, so it will affect *any/all* apps that don't express a preference. Works for Windows 10.16257 (and later)

![HKEY_CURRENT_USER\ConsoleVirtualTerminalLevel -- Registry setting to enable ANSI sequence processing on Windows 10.16257 and later](https://user-images.githubusercontent.com/5589855/36943005-11e41fe2-1f36-11e8-9401-7ac40d300f10.png)

**Comment by E3V3A** [comment]
@glenn-slayden 

> This changes the system-wide default for ANSI VT processing from opt-in to opt-out,..
> ...this registry setting would appear to be the only workaround for apps that can't be re-compiled to include a call to SetConsoleMode. 

Can this be understood as the setting overriding any `Get/SetConsoleMode` operations?
(I'm asking because I get very funny results running on W8.1 where reading the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` (using MinGW compiled ConoutMode.exe) tell me it is OFF, while CMD test seem to indicate that it is ON (ok), while PowerShell is behaving irregularly across versions and returns different results, most of the time no color (for obvious reasons), but returning the escape sequencies, so when run via alternative conhost terminal it works...sometimes.

**Comment by zadjii-msft** [comment]
`ENABLE_VIRTUAL_TERMINAL_PROCESSING` Is only available on Windows 10 and up, in the v2 console.

The v1 console (the only version of the console available on pre-Windows 10) *probably* doesn't ignore flags that it doesn't support, so you could probably SetConsoleMode with `ENABLE_VIRTUAL_TERMINAL_PROCESSING` and then have it roundtrip to your next call to GetConsoleMode, but that doesn't imply that it's actually enabled. I know that the v2 console will ignore flags it doesn't understand, but I'm not positive about the v1.

**Comment by koryovip** [comment]
jna is ok too

**Comment by quantum-hash** [comment]
### Use this to enable terminal colors by updating the registry using Java :
```
public class ENABLE_TERMINAL_COLORS_JAVA {
  public static void main(String[] args)
  {
    try { // TO ENABLE
      Process process = Runtime.getRuntime().exec("reg add HKEY_CURRENT_USER\\Console /v VirtualTerminalLevel /t REG_DWORD /d 0x00000001 /f");
    }
    // TO DISABLE (Put it back to its original value) --> try {Process process = Runtime.getRuntime().exec("reg add HKEY_CURRENT_USER\\Console /v VirtualTerminalLevel /t REG_DWORD /d 0x00000000 /f");}
    catch (Exception e) {
      e.printStackTrace(); // Catch and display exception
    }
  }
} 
```

> [![Click to follow - GitHub Followers](https://img.shields.io/github/followers/quantum-hash?label=Click%20to%20follow%20-%20GitHub%20Followers%20:&style=flat&color=success)](https://github.com/quantum-hash)
